### PR TITLE
URI and Config builder refactor

### PR DESCRIPTION
**Pull request by nerfZael**
## Summary of most important changes
- Builder refactor adds backwards compatibility, extensibility and ease of use
- Default config split into niche config bundles
- Use `uri!` macro for URI literals, use `Uri::try_from("..")` and handle errors when converting runtime strings.
- Use `HashMap::from` with builder to improve readability.
- `ResolutionResultCacheResolver` can now skip cache based on the resolving URI

## By-part breakdown
- **Config builder**
    - Core crate now has:
    ```rust
    pub trait ClientConfigBuilder {
        fn build(self) -> ClientConfig;
    }
    ```
    - `BuilderConfig` is now called `PolywrapClientConfig` as it is a Polywrap opinionated `ClientConfig`
    - builder crate has a `PolywrapClientConfigBuilder` trait which is an opinionated builder (hence prefixed by `Polywrap`). It extends the `ClientConfigBuilder` trait (from core). The `PolywrapClientConfig` implements `PolywrapClientConfigBuilder`.
    - Moved logic for creating a `StaticResolver` into a method on `PolywrapClientConfig`
    - `PolywrapClientConfig` can be created as empty, with `::new()` or with defaults as `::default()`. The two are currently the same, but might not be in the future.
    - `PolywrapClientConfig` can now be turned into a `ClientConfig` with a simple `.into()`. This is more idiomatic and easier for a dev to do since it doesn't require importing the `ClientConfigBuilder` trait.
    - `PolywrapBaseResolver` is a resolver used by the builder. It can be created with `::new(opt)` by specifying options or with `::default()`. This can even be used by developers without the builder for easy default resolver creation.
    - In places where `String` has been used instead of `Uri` (like the config builder) we are now using `Uri`.
    - In places where we were using `Vec` we are now using `HashMap` (e.g redirects).
    - Most importantly all structs provide defaults, which can be used by developers to avoid breaking changes when new options are added. E.g. with `SomeOptions { ..Default::default() }` 
- **Default config**
    - Split bundles into System and Web3
    - General restructure improvements with things like `HashMap::from` for more readable code
- **URI**
    - `authority`, `path` and `uri` are no longer public. This prevents devs from changing them and creating invalid URIs. `authority` and `path` can be accesses by getters, and `uri` can be accesses with `.to_string()`.
    - removed `Uri::new` as that can panic while giving no indication of a panic being possible
    - `from_string` is renamed to a more precise name: `try_from_string`
    - added unsafe function `from_parts` to be used in contexts where parsing has already been done externally (e.g. macros or FFI)
- **URI macro**
    - added `polywrap_core_macros` crate (Re exported by `polywrap_client` crate for ease of use)
    - `uri!` macro instead of `Uri::try_from("..").unwrap()` or `Uri::new("..")` gives compile time correctness for uri string literals
- **Resolvers**
    - `ResolutionResultCacheResolver` now has an optional `skip_cache` handler that can be used to skip cache. It also derives `Default`, so more optional features can be added without breaking changes.

### CODE REVIEW

**Review by cbrzn — APPROVED** [review]
lgtm 🍏

File: packages/builder/src/polywrap_client_config.rs [hunk 1]
```diff
@@ -1,48 +1,72 @@
...
 };
+use polywrap_resolvers::static_resolver::{StaticResolver, StaticResolverLike};
+
+use crate::{PolywrapBaseResolver, PolywrapBaseResolverOptions, PolywrapClientConfigBuilder};
+
+#[derive(Default, Clone)]
+pub struct PolywrapClientConfig {
+    pub interfaces: Option<InterfaceImplementations>,
+    pub envs: Option<HashMap<Uri, Vec<u8>>>,
+    pub wrappers: Option<Vec<(Uri, Arc<dyn Wrapper>)>>,
```
Comment (dOrgJelli) [inline packages/builder/src/polywrap_client_config.rs:19]: NIT: we should start using "wrap" more in our codebases, instead of "wrappers"

**Review by dOrgJelli — APPROVED** [review]
Looks great!

File: packages/builder/src/polywrap_client_config.rs (hunk 1 above)
Comment (nerfZael) [reply packages/builder/src/polywrap_client_config.rs:19]: Good point, I'll do it in a follow up PR

**Pull request merged**